Suggest depicted country based on EXIF latitude/longitude

Would need to embed a rough database of countries/areas borders, or [a specialized library if that exists](http://softwarerecs.stackexchange.com/questions/28378/tiny-java-library-to-guess-country-from-latitude-longitude).

On using it, I am getting this error

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: fr.free.nrw.commons, PID: 29045
    java.lang.NoClassDefFoundError: Failed resolution of: Ljavax/imageio/ImageIO;
        at io.github.coordinates2country.Coordinates2Country.country(Coordinates2Country.java:62)
        at io.github.coordinates2country.Coordinates2Country.countryQID(Coordinates2Country.java:31)
        at fr.free.nrw.commons.repository.UploadRepository.getCountryDepictions(UploadRepository.java:283)
        at fr.free.nrw.commons.upload.depicts.DepictsPresenter.selectPlaceDepictions(DepictsPresenter.kt:97)
        at fr.free.nrw.commons.upload.depicts.DepictsFragment.onBecameVisible(DepictsFragment.java:121)
        at fr.free.nrw.commons.upload.UploadActivity.onPreviousButtonClicked(UploadActivity.java:492)
        at fr.free.nrw.commons.upload.categories.UploadCategoriesFragment.onPreviousButtonClicked(UploadCategoriesFragment.java:180)
        at fr.free.nrw.commons.upload.categories.UploadCategoriesFragment_ViewBinding$2.doClick(UploadCategoriesFragment_ViewBinding.java:52)
        at butterknife.internal.DebouncingOnClickListener.onClick(DebouncingOnClickListener.java:18)
        at android.view.View.performClick(View.java:7161)
        at android.view.View.performClickInternal(View.java:7133)
        at android.view.View.access$3500(View.java:804)
        at android.view.View$PerformClick.run(View.java:27416)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:241)
        at android.app.ActivityThread.main(ActivityThread.java:7582)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:941)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "javax.imageio.ImageIO" on path: DexPathList[[zip file "/data/app/fr.free.nrw.commons-9JSpJEveQGUVDzRbot0I-Q==/base.apk"],nativeLibraryDirectories=[/data/app/fr.free.nrw.commons-9JSpJEveQGUVDzRbot0I-Q==/lib/arm64, /data/app/fr.free.nrw.commons-9JSpJEveQGUVDzRbot0I-Q==/base.apk!/lib/arm64-v8a, /system/lib64, /product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:196)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at io.github.coordinates2country.Coordinates2Country.country(Coordinates2Country.java:62) 
        at io.github.coordinates2country.Coordinates2Country.countryQID(Coordinates2Country.java:31) 
        at fr.free.nrw.commons.repository.UploadRepository.getCountryDepictions(UploadRepository.java:283) 
        at fr.free.nrw.commons.upload.depicts.DepictsPresenter.selectPlaceDepictions(DepictsPresenter.kt:97) 
        at fr.free.nrw.commons.upload.depicts.DepictsFragment.onBecameVisible(DepictsFragment.java:121) 
        at fr.free.nrw.commons.upload.UploadActivity.onPreviousButtonClicked(UploadActivity.java:492) 
        at fr.free.nrw.commons.upload.categories.UploadCategoriesFragment.onPreviousButtonClicked(UploadCategoriesFragment.java:180) 
        at fr.free.nrw.commons.upload.categories.UploadCategoriesFragment_ViewBinding$2.doClick(UploadCategoriesFragment_ViewBinding.java:52) 
        at butterknife.internal.DebouncingOnClickListener.onClick(DebouncingOnClickListener.java:18) 
        at android.view.View.performClick(View.java:7161) 
        at android.view.View.performClickInternal(View.java:7133) 
        at android.view.View.access$3500(View.java:804) 
        at android.view.View$PerformClick.run(View.java:27416) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:241) 
        at android.app.ActivityThread.main(ActivityThread.java:7582) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:941) 

```
Maybe the `java.awt` and `javax.imageio` isn't supported on Android

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -148,6 +148,8 @@ dependencies {
     //Glide
     implementation 'com.github.bumptech.glide:glide:4.12.0'
     annotationProcessor 'com.github.bumptech.glide:compiler:4.12.0'
+
+    implementation("io.github.coordinates2country:coordinates2country-android:1.2") {  exclude group: 'com.google.android', module: 'android' }
 }
 
 android {

--- a/app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java
+++ b/app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java
@@ -255,7 +255,7 @@ public class UploadRepository {
      */
 
     public Flowable<List<DepictedItem>> searchAllEntities(String query) {
-        return depictModel.searchAllEntities(query);
+        return depictModel.searchAllEntities(query, this);
     }
 
     /**
@@ -265,14 +265,14 @@ public class UploadRepository {
      * @return a single that provides the depictions
      */
     public Single<List<DepictedItem>> getPlaceDepictions() {
-        final Set<Place> places = new HashSet<>();
+        final Set<String> qids = new HashSet<>();
         for (final UploadItem item : getUploads()) {
             final Place place = item.getPlace();
             if (place != null) {
-                places.add(place);
+                qids.add(place.getWikiDataEntityId());
             }
         }
-        return depictModel.getPlaceDepictions(new ArrayList<>(places));
+        return depictModel.getPlaceDepictions(new ArrayList<>(qids));
     }
 
     /**

--- a/app/src/main/java/fr/free/nrw/commons/upload/structure/depictions/DepictModel.kt
+++ b/app/src/main/java/fr/free/nrw/commons/upload/structure/depictions/DepictModel.kt
@@ -31,14 +31,14 @@ class DepictModel @Inject constructor(private val depictsClient: DepictsClient) 
         return if (query.isBlank()) {
             nearbyPlaces.switchMap { places: List<Place> ->
                 val qids = mutableSetOf<String>()
-                repository.uploads.forEach {
-                    if(it.gpsCoords != null && it.gpsCoords.imageCoordsExists) {
-                        qids.add("Q"+ Coordinates2Country.countryQID(it.gpsCoords.decLatitude,
-                            it.gpsCoords.decLongitude))
-                    }
-                }
                 for(place in  places) {
                     place.wikiDataEntityId?.let { qids.add(it) }
+                }
+                repository.uploads.forEach { item ->
+                    if(item.gpsCoords != null && item.gpsCoords.imageCoordsExists) {
+                        Coordinates2Country.countryQID(item.gpsCoords.decLatitude,
+                            item.gpsCoords.decLongitude)?.let { qids.add("Q$it") }
+                    }
                 }
                 getPlaceDepictions(ArrayList(qids)).toFlowable()
             }

--- a/app/src/main/java/fr/free/nrw/commons/upload/structure/depictions/DepictModel.kt
+++ b/app/src/main/java/fr/free/nrw/commons/upload/structure/depictions/DepictModel.kt
@@ -2,6 +2,8 @@ package fr.free.nrw.commons.upload.structure.depictions
 
 import fr.free.nrw.commons.explore.depictions.DepictsClient
 import fr.free.nrw.commons.nearby.Place
+import fr.free.nrw.commons.repository.UploadRepository
+import io.github.coordinates2country.Coordinates2Country
 import io.reactivex.Flowable
 import io.reactivex.Observable
 import io.reactivex.Single
@@ -25,27 +27,37 @@ class DepictModel @Inject constructor(private val depictsClient: DepictsClient) 
     /**
      * Search for depictions
      */
-    fun searchAllEntities(query: String): Flowable<List<DepictedItem>> {
-        return if (query.isBlank())
+    fun searchAllEntities(query: String, repository: UploadRepository): Flowable<List<DepictedItem>> {
+        return if (query.isBlank()) {
             nearbyPlaces.switchMap { places: List<Place> ->
-                getPlaceDepictions(places).toFlowable()
+                val qids = mutableSetOf<String>()
+                repository.uploads.forEach {
+                    if(it.gpsCoords != null && it.gpsCoords.imageCoordsExists) {
+                        qids.add("Q"+ Coordinates2Country.countryQID(it.gpsCoords.decLatitude,
+                            it.gpsCoords.decLongitude))
+                    }
+                }
+                for(place in  places) {
+                    place.wikiDataEntityId?.let { qids.add(it) }
+                }
+                getPlaceDepictions(ArrayList(qids)).toFlowable()
             }
-        else
+        } else
             networkItems(query)
     }
 
     /**
-     * Provides [DepictedItem] instances via a [Single] for a given list of [Place], providing an
-     * empty list if no places are provided or if there is an error
+     * Provides [DepictedItem] instances via a [Single] for a given list of ids, providing an
+     * empty list if no places/country are provided or if there is an error
      */
-    fun getPlaceDepictions(places: List<Place>): Single<List<DepictedItem>> =
-        places.toIds().let { ids ->
+    fun getPlaceDepictions(qids: List<String>): Single<List<DepictedItem>> =
+        qids.toIds().let { ids ->
             if (ids.isNotEmpty())
                 depictsClient.getEntities(ids)
                     .map{
                         it.entities()
                             .values
-                            .mapIndexed { index, entity ->  DepictedItem(entity, places[index])}
+                            .mapIndexed { index, entity ->  DepictedItem(entity)}
                     }
                     .onErrorResumeWithEmptyList()
             else Single.just(emptyList())
@@ -75,7 +87,7 @@ class DepictModel @Inject constructor(private val depictsClient: DepictsClient) 
 
 }
 
-private fun List<Place>.toIds() = mapNotNull { it.wikiDataEntityId }.joinToString("|")
+private fun List<String>.toIds() = mapNotNull { it }.joinToString("|")
 
 private fun <T> Single<List<T>>.onErrorResumeWithEmptyList() = onErrorResumeNext { t: Throwable ->
     Single.just(emptyList<T>()).also { Timber.e(t) }

